### PR TITLE
[SYCL][ESIMD] Implement IR pass to lower C++ ESIMD intrinsics.

### PR DESCRIPTION
**Pull request by kbobrovs**
DO NOT REVIEW THE FIRST COMMIT (covered by https://github.com/intel/llvm/pull/1880)

The pass transforms *__esimd_* Itanium - mangled C++ intrinsics to
genx.*style parseable by the ESIMD - capable SPIRV translator.

This pass depends on the vc-intrinsics repo (GenXIntrinsics component), which is not part of the build yet - in progress.

Authors:
Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>
Gang Chen <gang.y.chen@intel.com>
Wei Pan
Denis Bakhvalov <denis.bakhvalov@intel.com>
Anton Sidorenko <anton.sidorenko@intel.com>
Kaiyu Chen <kai.yu.chen@intel.com>
Pratik Ashar <pratik.j.ashar@intel.com>

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 1]
```diff
@@ -0,0 +1,1316 @@
...
+    // arguments, each string describing argument type in OpenCL.
+    // required for running on top of OpenCL runtime.
+    SmallVector<Metadata *, 8> ArgTypeDescs;
+
+    auto *KernelArgTypes = F.getMetadata("kernel_arg_type");
+    unsigned Idx = 0;
+
+    // Iterate argument list to gather argument kinds and generate argument
+    // descriptors.
+    for (auto AI = F.arg_begin(), AE = F.arg_end(); AI != AE; ++AI) {
```
Comment (againull) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1137]: Nice to have:  range-based for loop over F.args().

**Comment by kbobrovs** [comment]
> Overall looks good to me, just minor comments. And currently build is failing with this error:
> 
> ```
> CMake Error at cmake/modules/AddLLVM.cmake:646 (add_dependencies):
>   The dependency target "LLVMGenXIntrinsics" of target "LLVMSYCLLowerIR" does
>   not exist.
> ```
> 
> Coule you please fix this.

Right. As I wrote in the description, this is pending (other people are working on this already): 
> This pass depends on the vc-intrinsics repo (GenXIntrinsics component), which is not part of the
> build yet - in progress.

File: llvm/include/llvm/SYCLLowerIR/LowerESIMD.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
+//===-- LowerESIMD.cpp - lower Explicit SIMD (ESIMD) constructs -----------===//
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/include/llvm/SYCLLowerIR/LowerESIMD.h:1]: ok

File: llvm/test/SYCLLowerIR/esimd_lower_intrins.ll [hunk 3]
```diff
@@ -0,0 +1,198 @@
+; This test checks C++ ESIMD intrinsics lowering to "@llvm.genx.*" form
+; consumable by the CM back-end.
+;
+; RUN: opt < %s -LowerESIMD -S | FileCheck %s
+
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64"
+target triple = "spir64-unknown-windows-sycldevice"
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/test/SYCLLowerIR/esimd_lower_intrins.ll:7]: ok

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 4]
```diff
@@ -0,0 +1,1316 @@
...
+        Kinds,
+        getMD(llvm::ConstantInt::getNullValue(I32Ty)), // SLM size in bytes
+        getMD(llvm::ConstantInt::getNullValue(I32Ty)), // arg offsets
+        IOKinds,
+        ArgDescs};
+
+    // Add this kernel to the root.
+    Kernels->addOperand(MDNode::get(Ctx, MDArgs));
+    F.addFnAttr("oclrt", "1");
+    // F.setDLLStorageClass(llvm::GlobalValue::DLLExportStorageClass);
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1180]: ok

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp (hunk 1 above)
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1137]: ok

**Comment by againull** [comment]
> > Overall looks good to me, just minor comments. And currently build is failing with this error:
> > ```
> > CMake Error at cmake/modules/AddLLVM.cmake:646 (add_dependencies):
> >   The dependency target "LLVMGenXIntrinsics" of target "LLVMSYCLLowerIR" does
> >   not exist.
> > ```
> > 
> > 
> > Coule you please fix this.
> 
> Right. As I wrote in the description, this is pending (other people are working on this already):
> 
> > This pass depends on the vc-intrinsics repo (GenXIntrinsics component), which is not part of the
> > build yet - in progress.

Oh, I see, thanks!

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 5]
```diff
@@ -0,0 +1,1313 @@
...
+
+    if (Name1.startswith(SPIRV_INTRIN_PREF)) {
+      auto SpirvPrefLen = StringRef(SPIRV_INTRIN_PREF).size();
+      StringRef SpirvIntrName = Name1.substr(SpirvPrefLen);
+      translateSpirvIntrinsic(CI, SpirvIntrName, ESIMDToErases);
+      // For now: if no match, just let it go untranslated.
+      continue;
+    }
+
+    if ((Name1.size() == 0) || !Name1.startswith(ESIMD_INTRIN_PREF1))
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1297]:
```suggestion
    if (Name1.empty() || !Name1.startswith(ESIMD_INTRIN_PREF1))
```

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 6]
```diff
@@ -0,0 +1,1313 @@
...
+    if (!CI || CI->isIndirectCall() || !(Callee = CI->getCalledFunction()))
+      continue;
+    StringRef Name = Callee->getName();
+
+    // See if the Name represents an ESIMD intrinsic and demangle only if it
+    // does.
+    if (!Name.startswith(ESIMD_INTRIN_PREF0))
+      continue;
+    // now skip the digits
+    StringRef Name1 = Name.substr(std::strlen(ESIMD_INTRIN_PREF0));
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1245]: I'm also not sure why do you need `Name1` here if you can just re-assign result of `substr` or `drop_while` back to `Name`

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 7]
```diff
@@ -0,0 +1,1313 @@
...
+      .Case("unsigned short", IntegerType::getInt16Ty(Ctx))
+      .Case("int", IntegerType::getInt32Ty(Ctx))
+      .Case("unsigned int", IntegerType::getInt32Ty(Ctx))
+      .Case("unsigned", IntegerType::getInt32Ty(Ctx))
+      .Case("unsigned long long", IntegerType::getInt64Ty(Ctx))
+      .Case("long long", IntegerType::getInt64Ty(Ctx))
+      .Case("float", IntegerType::getFloatTy(Ctx))
+      .Case("double", IntegerType::getDoubleTy(Ctx))
+      .Case("void", IntegerType::getVoidTy(Ctx))
+      .Case("", nullptr)
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:443]:
```suggestion
```

"default" branch anyway will return `nullptr`

**Comment by kbobrovs** [comment]
> Have you considered using existing clang infrastructure for built-ins?

Yes we have. Big difference with e.g. AMD GPU is that we compile with SPIRV target, not with Gen-specific target. SPIRV target can't have Gen-specific intrinsics enabled easily and non-intrusively. Another difference is that there is no Gen back-end, and no Gen TargetMachine, which also complicates things.

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp (hunk 7 above)
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:443]: ok

**Comment by AlexeySachkov** [comment]
> Big difference with e.g. AMD GPU is that we compile with SPIRV target, not with Gen-specific target. SPIRV target can't have Gen-specific intrinsics enabled easily and non-intrusively. Another difference is that there is no Gen back-end, and no Gen TargetMachine, which also complicates things.

AFAIK, you can have not only backend-specific built-ins, but also lang-specific or even lib-specific built-ins. Also, I don't fully understand why do you need actual backend or `TargetMachine` to add new intrinsics and emit them in LLVM IR (for native code, yes, some BE is required, but here we are talking about LLVM IR only)

**Comment by kbobrovs** [comment]
> AFAIK, you can have not only backend-specific built-ins, but also lang-specific or even lib-specific built-ins. 

Can you please please be more specific about your suggestion and why it is better than what's implemented? We can discuss this as a potential future alternative and add a TODO (if we agree that it is something worth doing). So far:
- this is really back-end specific, not lang or lib-specific intrinsics
- there are tons of combinations of template parameters which would require tons of new built-in definitions and lowering
- we use SPIRV target, which is device neutral, but need to represent and emit device-specific intrinsics, I don't think there is a prescribed recipe for this in LLVM, and I believe the solution implemented fits the purpose. But I'll be definitely interested in good alternative suggestions.

> Also, I don't fully understand why do you need actual backend or TargetMachine to add new intrinsics and emit them in LLVM IR (for native code, yes, some BE is required, but here we are talking about LLVM IR only)

What I said is that absence of TargetMachine will (well, is likely to) complicate implementation of this approach. You showed x86 and AMD as an example, but similar approach for Gen won't work w/o target machine.

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 8]
```diff
@@ -0,0 +1,1313 @@
...
+  });
+  translateSpirvIntr("GlobalSize", [](CallInst &CI, StringRef Suff) {
+    // GlobalSize = WorkGroupSize * NumWorkGroups
+    Instruction *WGSizeI =
+        generateVectorGenXForSpirv(CI, Suff, "local.size.v3i32", "wgsize.");
+    Instruction *NumWGI = generateVectorGenXForSpirv(
+        CI, Suff, "group.count.v3i32", "group_count.");
+    return BinaryOperator::CreateMul(WGSizeI, NumWGI, "mul", &CI);
+  });
+  // TODO: Support GlobalOffset SPIRV intrinsics
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:864]: Yes

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp (hunk 6 above)
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1245]: I'll see if I can simplify

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 9]
```diff
@@ -0,0 +1,1313 @@
...
+    // does.
+    if (!Name.startswith(ESIMD_INTRIN_PREF0))
+      continue;
+    // now skip the digits
+    StringRef Name1 = Name.substr(std::strlen(ESIMD_INTRIN_PREF0));
+    Name1 = Name1.drop_while([](char C) { return std::isdigit(C); });
+
+    // process ESIMD builtins that go through special handling instead of
+    // the translation procedure
+    if (Name1.startswith("cl4sycl5intel3gpu8slm_init")) {
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1248]: Yes, as quite a few places in other components - FE. But this is a good catch: the problem that I see here is that this intrinsic is not covered by a test case in the test below.
@cmc-rep, I think you added SLM support - do you have a test case for that?

[the remark this answers is not in the document]
Comment (cmc-rep) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1248]: We have slm_barrier.cpp and histogram_slm_256.cpp both testing SLM functions

**Comment by AlexeySachkov** [comment]
Please note that I'm not against going with the current approach, I'm just trying to understand whether we can implement the same in a way which is better aligned with existing stuff in clang & llvm (and potentially more reliable: dealing with mangler/demangler is evil)

> > Also, I don't fully understand why do you need actual backend or TargetMachine to add new intrinsics and emit them in LLVM IR (for native code, yes, some BE is required, but here we are talking about LLVM IR only)
> 
> What I said is that absence of TargetMachine will (well, is likely to) complicate implementation of this approach. You showed x86 and AMD as an example, but similar approach for Gen won't work w/o target machine.

Why so? I don't understand why emitting call to an intrinsic in `CGBuiltin.cpp` based on some descriptor from `Builtins.td` requires some TargetMachine and why it won't work for Gen.

> > AFAIK, you can have not only backend-specific built-ins, but also lang-specific or even lib-specific built-ins.
> 
> Can you please please be more specific about your suggestion and why it is better than what's implemented? 

Overall, I see that we could declare several new clang built-ins, which will be lowered into corresponding GenX intrinsics. Header files should be updated to call these new built-ins. Main advantages of that approach are the following:
- no dealing with demangling. I believe that relying on mangling might be fragile
- potentially better performance of the compiler itself: no need to demangle each EISMD-specific call
- this approach is more aligned with existing clang/llvm infrastructure

> We can discuss this as a potential future alternative and add a TODO (if we agree that it is something worth doing). So far:
> 
> * this is really back-end specific, not lang or lib-specific intrinsics

I don't see any problems with this

> * there are tons of combinations of template parameters which would require tons of new built-in definitions and lowering

I may miss something, so correct me if something is wrong, missing, or more complex than I think:

My understanding is that you are converting call to some function into intrinsic, based on some rules: for example:

> {"rdregion", {a(0), t(3), t(4), t(5), a(1), t(6)}, nk(-1)}}

Here first argument of the intrinsic is the first argument of a call, the second argument of the intrinsic is the third template argument of a called function, etc.

I think that absolutely the same can be expressed via existing clang built-ins infrastructure. If you know particular types which could be used in `a(0)` and `a(1)`, then you can declare a built-in with particular types (`t` is always just an integer value). `nk` shouldn't be documented in arguments at all - it will just be applied to intrinsic name when call to built-in is handled. There is no need in specific handling of template arguments or demangling anything: I suggest that this built-in is directly called from headers and all required template arguments are passed there as arguments.
If you don't know exact types which could be used in `a(0)` and `a(1)` or there are several possibilities, then you just need a built-in with custom type-checking.

> * we use SPIRV target, which is device neutral, but need to represent and emit device-specific intrinsics, I don't think there is a prescribed recipe for this in LLVM, 

There is no way to represent any arbitrary LLVM intrinsic in SPIR-V: this is not LLVM IR representation or generation problem, this is SPIR-V specification design. If you want to have more LLVM IR intrinsics represented in SPIR-V, then you need to have a SPIR-V spec updated or SPIR-V spec extension prepared or SPIR-V extended instruction set prepared.

> and I believe the solution implemented fits the purpose. But I'll be definitely interested in good alternative suggestions.

I think this patch implements a conversion from DPC++ methods/functions into LLVM IR intrinsics, but it has nothing to do with SPIR-V.

File: llvm/lib/SYCLLowerIR/LowerESIMD.cpp [hunk 10]
```diff
@@ -1027,7 +1026,7 @@ static void translateESIMDIntrinsicCall(CallInst &CI) {
   Function *F = CI.getCalledFunction();
   StringRef MnglName = F->getName();
   const char *MnglNameCStr = MnglName.data();
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1028]: This variable is probably became unused

Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1028]: > understand whether we can implement the same in a way which is better aligned with existing stuff in clang & llvm (and potentially more reliable: dealing with mangler/demangler is evil)

The word "better" and "evil" are highly subjective in this case :) I have the opposite opinion, for example. That's OK. I would suggest that you file an issue after we merge this patch and put your design sketches there. We definitely need more people to discuss what's the best here if there are doubts. Does this sound good to you?

> There is no way to represent any arbitrary LLVM intrinsic in SPIR-V: 

As far as I can see `--spirv-allow-unknown-intrinsics` serves pretty well for this purpose. What exactly do you mean?

> I think this patch implements a conversion from DPC++ methods/functions into LLVM IR intrinsics

right

>, but it has nothing to do with SPIR-V.

It has. This is done for the SPIRV target which means I can't use device-specific built-ins like AMDGPU built-ins, which are enabled for specific target. So SPIRV is essential here.

Actually my main doubt/objection about using clang is not really complications related to hacking built-ins mechanism for our purposes. It is the need to introduce a lot of new intrinsics derived from combinations of element type, vector length and some other parameters. Not only I'd have to introduce those, I'd also have to write lowering for each, of course.

Overall, I think this kind of debate should better happen in the dedicated issue.

Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1028]: > Why so? I don't understand why emitting call to an intrinsic in CGBuiltin.cpp based on some descriptor from Builtins.td requires some TargetMachine and why it won't work for Gen.

Do you mean Builtins.def? This is target-neutral list of built-ins. It is not a good place for Gen intrinsics. Also I don't think e.g. ARM built-ins can be used with targets other than ARM.

Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/LowerESIMD.cpp:1028]: > This variable is probably became unused

yep, I haven't yet pushed the fixed variant, just committed you suggestions. Will do shortly

**Comment by kbobrovs** [comment]
Had to rebase and force-push.

**Comment by kbobrovs** [comment]
Had to rebase again to fix another problem in ItaniumDemangler.h.
@AlexeySachkov - please review and approve if no more objections.
@againull - could you please re-submit approval.

File: llvm/include/llvm/Demangle/ItaniumDemangle.h [hunk 11]
```diff
@@ -445,6 +445,14 @@ class EnableIfAttr : public Node {
   }
 };
 
+#ifdef _MSC_VER
```
Comment (bader) [inline llvm/include/llvm/Demangle/ItaniumDemangle.h:448]: Hm... is this a GitHub bug or this code will be added again?
It should be already in place - https://github.com/intel/llvm/pull/2021.

Comment (kbobrovs) [reply llvm/include/llvm/Demangle/ItaniumDemangle.h:448]: This is to make compilation and testing for this PR pass. I will rebase and this will go away.

**Pull request merged**